// @fluentui/react-component-ref/dist/dts/test/fixtures.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ForwardedRef = exports.CompositeClass = exports.DOMClass = exports.CompositeFunction = exports.DOMFunction = void 0;
var React = require("react");
var DOMFunction = function (props) { return React.createElement("div", __assign({}, props, { id: "node" })); };
exports.DOMFunction = DOMFunction;
var CompositeFunction = function (props) { return React.createElement(exports.DOMFunction, __assign({}, props)); };
exports.CompositeFunction = CompositeFunction;
var DOMClass = /** @class */ (function (_super) {
    __extends(DOMClass, _super);
    function DOMClass() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DOMClass.prototype.render = function () {
        return React.createElement("div", __assign({}, this.props, { id: "node" }));
    };
    return DOMClass;
}(React.Component));
exports.DOMClass = DOMClass;
var CompositeClass = /** @class */ (function (_super) {
    __extends(CompositeClass, _super);
    function CompositeClass() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    CompositeClass.prototype.render = function () {
        return React.createElement(DOMClass, __assign({}, this.props));
    };
    return CompositeClass;
}(React.Component));
exports.CompositeClass = CompositeClass;
exports.ForwardedRef = React.forwardRef(function (props, ref) { return (React.createElement("div", null,
    React.createElement("button", __assign({}, props, { ref: ref })))); });
